processor/src/host/debug.rs

```diff
@@ -39,14 +82,173 @@ fn print_vm_stack(clk: u32, stack: Vec<Felt>, n: usize) {
     // un-printed items
     let i = num_items - 1;
     if num_items == stack.len() {
-        println!("└── {i:>2}: {}", stack[i]);
+        println!("└── {i:>2}: {}\n", stack[i]);
     } else {
         println!("├── {i:>2}: {}", stack[i]);
-        println!("└── ({} more items)", stack.len() - num_items);
+        println!("└── ({} more items)\n", stack.len() - num_items);
+    }
+}
+
+/// Prints the whole memory state at the cycle `clk` in context `ctx`.
+#[cfg(feature = "std")]
+fn print_mem_all(clk: u32, ctx: u32, mem: Vec<(u64, Word)>) {
+    println!("Memory state before step {clk} for the context {ctx}:");
+
+    // calculate the proper padding for the memory addresses
+    let padding = mem.iter().fold(0, |max, value| value.0.to_string().len().max(max));
+
+    // print the main part of the memory (wihtout the last value)
+    for (addr, value) in mem.iter().take(mem.len() - 1) {
+        println!(
+            "├── {addr:>width$}: [{}, {}, {}, {}]",
+            value[0],
+            value[1],
+            value[2],
+            value[3],
+            width = padding
+        );
```

Comment:
I would probably create a separate function to encapsulate this logic. Maybe something like:
```Rust
fn print_mem_address(clk: u32, ctx: u32, addr: u32, value: Option<Word>, is_last: bool)
```
Also, I wonder if we can just apply universal padding. Since memory addresses are 32 bits, we should fit into 8 hex characters (it is common to print memory address in hex).

We could probably have a similar function for locals (e.g., `print_loc_address`) with the main difference being that index of local should be printed out in decimal (rather than hex). Or we could combine the two functions and have an extra parameter to differentiate between the two.